### PR TITLE
New tutorial - Transparent Proxy and Filtering on K8S

### PR DESCRIPTION
**Pull request by danisla**
New tutorial showing how to deploy and use a transparent proxy on Kubernetes to filter and intercept egress pod traffic. 

Follow-on tutorial that uses initializers in #252 

- [x] Tech writing / styleguide review - @steveperry-53
- [x] Subject matter review - @ahmetb
- [x] Formatting review for publishing - @jmdobry

### CODE REVIEW

**Comment by hegemonic** [comment]
@JimTravis, can you assign a new tech-writer reviewer? I'm busy getting ready for a launch right now.

**Comment by steveperry-53** [comment]
@danisla I'll take a close look at this later today. For starters, see whether the changes suggested by @ccarpentiere in https://github.com/GoogleCloudPlatform/community/pull/252 apply here.

**Comment by steveperry-53** [comment]
@danisla The sed command is not working for me.

$ cd charts/tproxy/
$ sed -i '' -e 's/418/500/g' config/mitm-script.py
sed: can't read : No such file or directory

$ cd config
$ sed -i '' -e 's/418/500/g' mitm-script.py
sed: can't read : No such file or directory

File: tutorials/transparent-proxy-and-filtering-on-k8s/index.md [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+There are many application environments and scenarios where you may want to filter and intercept all HTTP/S traffic out of a pod. Example use cases include isolating external access to specific HTTP or HTTPS paths, methods and in-flight altering of requests.
+
+An example of HTTPS interception would be filtering access to Google Cloud Storage buckets. Most proxy solutions do not support HTTPS inspection so the filtering that can be done is limited at the IP/DNS level. For the case of a Cloud Storage bucket, the bucket name is embedded under the https://storage.googleapis.com/BUCKET_NAME/object URL so filtering is limited to all buckets under the storage.googleapis.com domain, not any specific bucket.
+
+Another example would be the in-flight modification of requests. If you wanted to transparently add request/response headers to requests between services for tracing purposes, this can be done without the service having to explicitly set them when the request is made.
+
+[Mitmproxy](https://mitmproxy.org/) is an Open Source tool that can be used to intercept and modify HTTP and HTTPS requests transparently using the Python scripting language.
```
Comment (steveperry-53) [inline tutorials/transparent-proxy-and-filtering-on-k8s/index.md]: Suggestion:
If you want to transparently add request/response headers to requests between services for tracing purposes, you can do that without the services having to explicitly set them.

File: tutorials/transparent-proxy-and-filtering-on-k8s/index.md [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        - name: ca-certs-debian
+        configMap:
+            name: tproxy-tproxy-root-certs
+            items:
+            - key: root-certs.crt
+            path: ca-certificates.crt
+
+
+3. Run the command below to get the status of the DaemonSet pods
+
```
Comment (steveperry-53) [inline tutorials/transparent-proxy-and-filtering-on-k8s/index.md]: You don't need, "Run the command below". You can just say, "Get the status of the DaemonSet pods:".

File: tutorials/transparent-proxy-and-filtering-on-k8s/index.md [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+![architecture diagram](./tproxy_diagram.png)
+
+**Figure 1.** transparent proxy architecture diagram
+
+## Objectives
+
+- Create a Kubernetes cluster with Google Container Engine
+- Deploy the tproxy and the tproxy-podwatch pods using Helm
+- Deploy example apps to test external access to a Cloud Storage bucket.
+
```
Comment (steveperry-53) [inline tutorials/transparent-proxy-and-filtering-on-k8s/index.md]: Periods at the ends of all list items.

File: tutorials/transparent-proxy-and-filtering-on-k8s/index.md [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+This tutorial uses billable components of GCP, including:
+
+- [Google Container Engine](https://cloud.google.com/container-engine/pricing)
+
+Use the [Pricing Calculator](https://cloud.google.com/products/calculator/#id=f52c2651-4b02-4da3-b8cd-fdbca6ad89a9) to estimate the costs for your environment. The resources used in this tutorial are approximately $2.39 per day.
+
+## Checkout the source repository
+
+1. Open [Cloud Shell](https://console.cloud.google.com/cloudshell)
+
```
Comment (steveperry-53) [inline tutorials/transparent-proxy-and-filtering-on-k8s/index.md]: I suggest opening Cloud Shell in a new tab. For me, it was hard to find my way back to the tutorial. {:target="_blank"}

File: tutorials/transparent-proxy-and-filtering-on-k8s/index.md [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+        10.12.1.41:36496: GET https://storage.googleapis.com/solutions-public-assets/adtech/dfp_networkimpressions.py HTTP/2.0
+                    << 200  (content missing)
+        10.12.1.41:36496: clientdisconnect
+
+    Notice that the proxy blocks the request to https://www.google.com with status code 418.
+
+## Customizing the mitmproxy Python script
+
+This tutorial uses a Python script filter traffic to a specific Cloud Storage bucket. The Python script is installed as a [ConfigMap](https://kubernetes.io/docs/tasks/configure-pod-container/configmap/) resource and mitmproxy can live reload the script when it changes without restarting.
+
```
Comment (steveperry-53) [inline tutorials/transparent-proxy-and-filtering-on-k8s/index.md]: script filter -> script to filter

**Comment by steveperry-53** [comment]
@danisla A few comments. Otherwise LGTM.

**Comment by jmdobry** [comment]
I'll do my review after @steveperry-53's comments are addressed and @ahmetb's review is complete.

**Comment by ahmetb** [comment]
@danisla do we really need this tutorial when #252 will work in just a couple of months? I think having both will confuse the users further. I will leave it up to you to push these both forward or choose one to go with. 

Alternatively, we can publish only this for now, and once v1.9 releases with Initializers, you can come back and edit the tutorial to do the task with the Initializers.

**Comment by danisla** [comment]
@steveperry-53 changes incorporated. Thanks.

@ahmetb, I want to keep both tutorials because even though 1.9 is a few months away, there are still a ton of 1.6 clusters out there so I think both are appropriate.

File: tutorials/transparent-proxy-and-filtering-on-k8s/index.md [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+        cd charts/tproxy
+        docker run --rm -v ${PWD}/certs/:/home/mitmproxy/.mitmproxy mitmproxy/mitmproxy >/dev/null 2>&1
+
+2. Install the chart:
+
+        helm install -n tproxy .
+
+    The output of this command shows you how to augment your deployments to use the init container and trusted certificate configmap volume. Example output below:
+
+        Add the init container spec below to your deployments:
```
Comment (ahmetb) [inline tutorials/transparent-proxy-and-filtering-on-k8s/index.md]: this renders as code on github make sure it doesn't render as such on website

File: tutorials/transparent-proxy-and-filtering-on-k8s/index.md [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+
+        initContainers:
+        - name: tproxy
+        image: docker.io/danisla/tproxy-sidecar:0.1.0
```
Comment (ahmetb) [inline tutorials/transparent-proxy-and-filtering-on-k8s/index.md]: field needs indentation

File: tutorials/transparent-proxy-and-filtering-on-k8s/index.md [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+            memory: 128Mi
+            requests:
+            cpu: 100m
+            memory: 64Mi
+
+        Add the volumes below to your deployments to use the trusted https tproxy:
+
+        volumes:
+        - name: ca-certs-debian
+        configMap:
```
Comment (ahmetb) [inline tutorials/transparent-proxy-and-filtering-on-k8s/index.md]: field needs indentation, not exactly under `n` of `name`

File: tutorials/transparent-proxy-and-filtering-on-k8s/index.md (hunk 6 above)
Comment (danisla) [reply tutorials/transparent-proxy-and-filtering-on-k8s/index.md]: I don't think we can distinguish between code and pre-formatted text in the tutorials. This block is meant to just show the output you see in the console.

File: tutorials/transparent-proxy-and-filtering-on-k8s/index.md (hunk 7 above)
Comment (danisla) [reply tutorials/transparent-proxy-and-filtering-on-k8s/index.md]: The actual helm chart outputs the correct indentation, this was just my markdown indenting mistake. I'll fix up the markdown.

**Comment by jmdobry** [comment]
@steveperry-53 and @ahmetb can you approve the PR?

**Comment by steveperry-53** [comment]
@danisla It looks like {: target="_blank"} is not working. My mistake for suggesting it. The Markdown processor is just sticking it outside the <a> element.

<a href="https://cloud.google.com/container-engine/pricing">Google Container Engine</a>{:target="_blank"}

There are two possible fixes:

* Just get rid of all the {: target="_blank"}s.
* Write those links as HTML <a> tags instead of using Markdown.

**Comment by steveperry-53** [comment]
LGTM except for the target="_blank" issue.

**Comment by jmdobry** [comment]
Community tutorials must be pure Markdown. HTML is stripped out. non-markdown formatting does not work.

**Comment by danisla** [comment]
@steveperry-53 ok I just removed the `{: target="_blank"}`s.

**Comment by danisla** [comment]
@jmdobry anything else blocking this from being merged?

**Comment by jmdobry** [comment]
Waiting for @ahmetb's green button approval.

**Comment by ahmetb** [comment]
Not sure why green button is required. but let's go.

**Comment by jmdobry** [comment]
Using the green approve button or writing LGTM on the PR after your comments are addressed let us know you're done with your review.

**Comment by jmdobry** [comment]
Published: https://cloud.google.com/community/tutorials/transparent-proxy-and-filtering-on-k8s

**Pull request merged**